rust/src/index/ann/sort.rs

```diff
@@ -0,0 +1,134 @@
+use std::cmp::Ordering;
+
+use arrow_schema::ArrowError;
+use rayon::prelude::*;
+use rand::{Rng, SeedableRng};
+
+/// Partition the `arr` for the first `topk` elements are the smallest
+pub fn find_topk<T: Send, F>(values: &mut [T], indices: &mut [u64], topk: usize, compare: &F) -> Result<(), ArrowError>
+    where F: Fn(&T, &T) -> Ordering + Sync {
+    let mut rng = rand::rngs::SmallRng::from_entropy();
+    assert!(topk > 0);
+    if values.len() <= topk {
+        return Ok(());
+    }
+    let mut low = 0;  // inclusive
+    let mut high = values.len();  // exclusive
+    // TODO handle worse case scenarios and fallback to nlog(n) sort
+    loop {
+        let pivot: usize = rng.gen_range(low..high);  // choose random pivot
+        let left = partition(values, indices, low, high, pivot, compare)?;
+        if left == topk {
+            return Ok(());
+        } else if left > topk {
+            high = left;
+        } else if left < topk {
+            low = left;
+        }
+    }
+}
+
+/// Returns the number of elements that are smaller than the pivot value
```

Comment:
This is fancy stuff.  isnt there some partition alrogithm / crates out there?

Reply:
rayon has something but the partitioning logic isn't a public function. I'd rather not maintain our own so if you see/know a good one lmk.

Reply:
also the other problem is it needs to "co-partition" the scores and row_id's together. So really we're looking for arg-topk